### PR TITLE
[DRAFT] #845 Added No. Series Where-Used

### LINKED ISSUE #845: [BC Idea]: 'Where-used' No. Series
### BC Idea Link

https://experience.dynamics.com/ideas/idea/?ideaid=d190df93-0e93-ed11-a76e-0003ff45e043

### Description

This issue is generated from the following idea on [aka.ms/bcideas](https://github.com/microsoft/BusinessCentralApps/issues/aka.ms/bcideas): ['Where-used' No. Series](https://experience.dynamics.com/ideas/idea/?ideaid=d190df93-0e93-ed11-a76e-0003ff45e043)

It would be great to be able to see where a specific number series is used. Idealy in the same way this can now be done from a general ledger account. It would be very useful to see which number series are obsolete, in stead of now going throug all the different setup pages to see if a no. series is used or not.

> Manually transfered from https://github.com/microsoft/BusinessCentralApps/issues/365

### I will provide the implementation for this BC Idea

- [X] I will provide the implementation for this BC Idea

Internal work item: [AB#460588](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/460588)

### PR DESCRIPTION
**Pull request by christianbraeunlich**
#### Summary 

This is a Draft PR. There's still a lot to do.

- [x] Adding tables manually (like in the G/L Account) or search for all "Table Relations Metadata" with "Related Table ID" + "Related Field No." set to No. Series / Code (this is my prefered way).
- [ ] Fix PageManagement so that the user can open the page from the Where-Used results.
- [ ] Obsolete
- [ ] Field with TableRelation = No. Series has value or not
- [ ] (Optional) Count
- [ ] automated tests

> Manually transfered from https://github.com/microsoft/BusinessCentralApps/pull/427

#### Work Item(s) 
Fixes #845

### CODE REVIEW

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    trigger OnRun()
+    begin
+    end;
+
+    var
+        TempNoSeriesWhereUsed: Record "No. Series Where-Used" temporary;
+        NextEntryNo: Integer;
+
+    procedure ShowSetupForm(NoSeriesWhereUsed: Record "No. Series Where-Used")
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al:18]: "Form"? :-)

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        CheckNoSeries(NoSeriesCode);
+        ShowNoSeriesWhereUsed();
+    end;
+
+    local procedure ShowNoSeriesWhereUsed()
+    begin
+        OnBeforeShowNoSeriesWhereUsed(TempNoSeriesWhereUsed);
+
+        TempNoSeriesWhereUsed.SetCurrentKey("Table Name");
+        PAGE.RunModal(0, TempNoSeriesWhereUsed);
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
        Page.RunModal(0, TempNoSeriesWhereUsed);
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        OnAfterCheckNoSeries(TempNoSeriesWhereUsed, NoSeriesCode);
+    end;
+
+    local procedure FillTableBuffer(var TableBuffer: Record "Integer"): Boolean
+    var
+        NoSeries: Record "No. Series";
+        TableRelationsMetadata: Record "Table Relations Metadata";
+    begin
+        TableRelationsMetadata.SetLoadFields("Related Table ID", "Related Field No.");
+        TableRelationsMetadata.SetRange("Related Table ID", DATABASE::"No. Series");
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
        TableRelationsMetadata.SetRange("Related Table ID", Database::"No. Series");
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        "Field": Record "Field";
+        RecRef: RecordRef;
+    begin
+        RecRef.Open(TableID);
+        TempNoSeriesWhereUsed.Init();
+        TempNoSeriesWhereUsed."Table ID" := TableID;
+        TempNoSeriesWhereUsed."Table Name" := RecRef.Caption;
+
+        TableRelationsMetadata.SetRange("Table ID", TableID);
+        TableRelationsMetadata.SetRange("Related Table ID", DATABASE::"No. Series");
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
        TableRelationsMetadata.SetRange("Related Table ID", Database::"No. Series");
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                    TempNoSeriesWhereUsed."Key 1" := CopyStr(Format(FieldRef.Value), 1, MaxStrLen(TempNoSeriesWhereUsed."Key 1"));
+                2:
+                    TempNoSeriesWhereUsed."Key 2" := CopyStr(Format(FieldRef.Value), 1, MaxStrLen(TempNoSeriesWhereUsed."Key 2"));
+                3:
+                    TempNoSeriesWhereUsed."Key 3" := CopyStr(Format(FieldRef.Value), 1, MaxStrLen(TempNoSeriesWhereUsed."Key 3"));
+                4:
+                    TempNoSeriesWhereUsed."Key 4" := CopyStr(Format(FieldRef.Value), 1, MaxStrLen(TempNoSeriesWhereUsed."Key 4"));
+            end;
+        end;
+        NextEntryNo := NextEntryNo + 1;
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
        NextEntryNo += 1;
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+using System.Reflection;
+using System.Utilities;
+
+codeunit 1 "Calc. No. Series Where-Used"
+{
+
+    trigger OnRun()
+    begin
+    end;
+
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        if not TableBuffer.Get(TableID) then begin
+            TableBuffer.Number := TableID;
+            TableBuffer.Insert();
+        end;
+    end;
+
+    local procedure CheckTable(NoSeriesCode: Code[20]; TableID: Integer)
+    var
+        TableRelationsMetadata: Record "Table Relations Metadata";
+        "Field": Record "Field";
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
        Field: Record Field;
```

File: src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        TempNoSeriesWhereUsed.Line := '';
+        KeyRef := RecRef.KeyIndex(1);
+        for KeyFieldCount := 1 to KeyRef.FieldCount do begin
+            FieldRef := KeyRef.FieldIndex(KeyFieldCount);
+            FieldCaptionAndValue := StrSubstNo('%1=%2', FieldRef.Caption, FieldRef.Value);
+            if TempNoSeriesWhereUsed.Line = '' then
+                TempNoSeriesWhereUsed.Line := CopyStr(FieldCaptionAndValue, 1, MaxStrLen(TempNoSeriesWhereUsed.Line))
+            else
+                TempNoSeriesWhereUsed.Line :=
+                  CopyStr(TempNoSeriesWhereUsed.Line + ', ' + FieldCaptionAndValue, 1, MaxStrLen(TempNoSeriesWhereUsed.Line));
```
Comment (NKarolak) [inline src/Business Foundation/App/NoSeries/src/Setup/NoSeriesCalcWhereUsed.Codeunit.al]:
```suggestion
                    CopyStr(TempNoSeriesWhereUsed.Line + ', ' + FieldCaptionAndValue, 1, MaxStrLen(TempNoSeriesWhereUsed.Line));
```

**Review by NKarolak — COMMENTED** [review]
Just some super-minor findings along your way :-)
I did not check the business logic itself.

**Review by christianbraeunlich — COMMENTED** [review]
Apply suggestions from code review

**Comment by JesperSchulz** [comment]
Hi @christianbraeunlich,

we've been discussing your PR a little and we're now having some doubts about its validity. While it makes a ton of sense to be able to see where a No. Series is used, we're not sure that this feature should be build in the way it's being built right now. In AI, "where used" is a rather prominent capability, which must be generally available for all entity relationships. You will probably see some advancements in this area in the near future, and I hence think it makes sense to wait to add this capability, until we know what the future holds here.

If you agree, I'd suggest to close this DRAFT for now and pick it up again, once we know more.

What do you say?

**Comment by christianbraeunlich** [comment]
I completely 100% agree. Once I moved the code into the business foundation layer, everything felt out of place, and even before that, there's a need for more general functionality that would benefit all the other modules. So a WhereUsed module would make perfect sense. The same goes for the approval workflows imho, but I'll get to that in the otter PR.

**Comment by JesperSchulz** [comment]
Good to hear that  you've made the same observations 😊 Let's see what the future holds!

**Pull request closed without merging**